src/main/java/com/google/devtools/build/lib/syntax/MethodLibrary.java

```diff
@@ -102,6 +105,51 @@ private static Object findExtreme(
     }
   }
 
+  private static Object evalKey(
+      Object obj, Object key, Location loc, Environment env, FuncallExpression ast)
+      throws EvalException, InterruptedException {
+    if (key == Runtime.NONE) {
+      return obj;
+    } else if (key instanceof BuiltinCallable) {
+      return ((BuiltinCallable) key).call(Arrays.asList(obj), Collections.emptyMap(), ast, env);
+    } else if (key instanceof BaseFunction) {
+      return ((BaseFunction) key).call(Arrays.asList(obj), null, null, env);
+    }
+
+    throw new EvalException(loc, Printer.format("%r object is not callable", EvalUtils.getDataTypeName(key)));
+  }
+
+  private static class KeyComparator implements Comparator<Object> {
```

Comment:
Let's move this class declaration immediately prior to its sole point of use (inside 'sorted'). We can then simplify by letting lexical scope do its job. There should be no need for fields (other than e) or a constructor. They can just be local variables of the enclosing function.